src/hazelcore/semantics/Action.re

```diff
@@ -2483,6 +2499,41 @@ module Exp = {
     | MoveRight => syn_move(ctx, a, (zblock, ty, u_gen))
 
     /* Backspace & Delete */
+    // Handle 2 special cases for CommentLines
+    // Case 1
+    | Backspace when ZExp.is_begin_of_comment(zblock) =>
+      switch (prefix |> ListUtil.split_last) {
+      | Some((new_prefix, pre_zline)) =>
+        switch (pre_zline, zline) {
+        | (CommentLine(pre_comment), CursorL(_, CommentLine(comment))) =>
+          let new_zline =
+            ZExp.CursorL(
+              OnText(String.length(pre_comment)),
+              CommentLine(pre_comment ++ comment),
+            );
+          let new_ze = (new_prefix, new_zline, suffix);
+          Succeeded(SynDone((new_ze, ty, u_gen)));
+        | _ => Failed
+        }
+      | _ => Failed
+      }
+
+    // Case 2
+    | Delete when ZExp.is_end_of_comment(zblock) =>
+      switch (suffix, zline) {
+      | (
+          [CommentLine(post_comment), ...new_suffix],
+          CursorL(_, CommentLine(comment)),
+        ) =>
+        let new_zline =
+          ZExp.CursorL(
+            OnText(String.length(comment)),
+            CommentLine(comment ++ post_comment),
+          );
+        let new_ze = (prefix, new_zline, new_suffix);
+        Succeeded(SynDone((new_ze, ty, u_gen)));
+      | _ => Failed
+      }
```

Comment:
Rather than repeating the pattern matching on the prefix/suffix and zline (both in the `when` guard and in the rule clause), I would refactor the top level switch expression of `syn_perform_block` so that you pattern match on `(a, (ListUtil.split_last(prefix), zline, suffix))`. This will allow you to do the pattern matching just once.

Reply:
You'll need to repeat these cases in `ana_perform_block` too.

Reply:
Still need to address these comments